### PR TITLE
User referrals

### PR DESCRIPTION
**Pull request by lrojas94**
**Sample URL:** 
Code-based: http://localhost:3000/login?ref_code=LRO-RVL
Source-based; http://localhost:3000/login?ref_source=my-source

image
image
image

@JustMaier tagging you here in case you wanna take a look/comment on the approach.

### CODE REVIEW

File: src/server/schema/user.schema.ts [hunk 1]
```diff
@@ -109,3 +111,6 @@ export type ToggleUserArticleEngagementsInput = z.infer<typeof toggleUserArticle
 export const toggleUserArticleEngagementSchema = getUserArticleEngagementsSchema.extend({
   articleId: z.number(),
 });
+
+export const userByReferralCodeSchema = z.object({ userReferralCode: z.string().min(3) });
```
Comment (manuelurenah) [inline src/server/schema/user.schema.ts:115]: You did it here 😛

File: src/components/OnboardingModal/OnboardingModal.tsx [hunk 2]
```diff
@@ -213,6 +221,38 @@ export default function OnboardingModal() {
...
+                      size="lg"
+                      name="userReferralCode"
+                      label="Referral Code"
+                      type="text"
+                      clearable={false}
+                      error={
+                        userReferralCode && userReferralCode.length <= 3
+                          ? 'Referral codes must be longer than 3 characters'
+                          : undefined
+                      }
```
[the remark this answers is not in the document]
Comment (lrojas94) [reply src/components/OnboardingModal/OnboardingModal.tsx:235]: No :( Zod doesn't allow partial + min

File: src/server/schema/user.schema.ts (hunk 1 above)
Comment (lrojas94) [reply src/server/schema/user.schema.ts:115]: cause there it's not optional :D

**Review by manuelurenah — APPROVED** [review]
🍕

**Comment by lrojas94** [comment]
@JustMaier this is good for review. 
One nit. Right now, I'm allowing a referral (via source) to be updated with a Ref. code from a user (Mainly in the case that a user happens to get both source + code). I don't know if this'll ever make sense tbh, so lmk. I jump a few hoops to make this possible and we might not even need it.

File: prisma/schema.prisma [hunk 3]
```diff
@@ -66,11 +66,12 @@ model UserReferral {
...
-  code   String  @unique
-  note   String?
+  id        Int       @id @default(autoincrement())
+  userId    Int
+  user      User      @relation(fields: [userId], references: [id], onDelete: Cascade)
+  code      String    @unique
+  note      String?
+  deletedAt DateTime?
 
   referees UserReferral[]
```
Comment (manuelurenah) [inline prisma/schema.prisma:76]: I learned from Koen that we can have an index on `userId` to make querying these faster 🤓 
```suggestion
  referees UserReferral[]
  
  @@index([userId], type: Hash)
```

File: src/components/Referrals/ReferralsProvider.tsx [hunk 4]
```diff
@@ -33,21 +33,23 @@ export const ReferralsProvider = ({ children }: { children: React.ReactNode }) =
...
   useEffect(() => {
-    if (result.success) {
+    if (result.success && !user?.referral) {
       const { ref_id, ref_source, ref_code } = result.data;
       const { code, source } = referrals;
-      const expirationDate = dayjs().add(7, 'day').toDate();
+      const expirationDate = dayjs().add(5, 'day').toDate();
 
       if (ref_id && ref_id !== code) {
+        console.log('Cookie set');
```
Comment (manuelurenah) [inline src/components/Referrals/ReferralsProvider.tsx:42]: Maybe leftover?
```suggestion
```

File: src/components/Referrals/ReferralsProvider.tsx [hunk 5]
```diff
@@ -33,21 +33,23 @@ export const ReferralsProvider = ({ children }: { children: React.ReactNode }) =
...
         setCookie('ref_code', ref_id, expirationDate);
         setCode(ref_id);
       }
       if (ref_code && ref_code !== code) {
-        setCookie('ref_code', ref_id, expirationDate);
+        console.log('Cookie set');
```
Comment (manuelurenah) [inline src/components/Referrals/ReferralsProvider.tsx:47]:
```suggestion
```

File: src/pages/api/auth/[...nextauth].ts [hunk 6]
```diff
@@ -10,14 +10,23 @@ import EmailProvider, { SendVerificationRequestParams } from 'next-auth/provider
...
+  getSessionUser,
+  updateAccountScope,
+} from '~/server/services/user.service';
 import { Tracker } from '~/server/clickhouse/client';
 import { NextApiRequest, NextApiResponse } from 'next';
 import { civitaiTokenCookieName, useSecureCookies } from '~/libs/auth';
 import { isDev } from '~/env/other';
 import { verificationEmail } from '~/server/email/templates';
 import blockedDomains from '~/server/utils/email-domain-blocklist.json';
+import { getCookie } from 'cookies-next';
```
Comment (manuelurenah) [inline src/pages/api/auth/[...nextauth].ts:29]: Is this even being used?

File: src/pages/api/auth/[...nextauth].ts [hunk 7]
```diff
@@ -153,14 +163,26 @@ export default async function auth(req: NextApiRequest, res: NextApiResponse) {
...
+
     if (context.isNewUser) {
       const tracker = new Tracker(req, res);
       await tracker.userActivity({
         type: 'Registration',
         targetUserId: parseInt(context.user.id),
       });
+
+      if (source) {
+        // Only source will be ser via the auth callback.
```
Comment (manuelurenah) [inline src/pages/api/auth/[...nextauth].ts:176]:
```suggestion
        // Only source will be set via the auth callback.
```

File: src/server/services/user.service.ts [hunk 8]
```diff
@@ -732,26 +732,44 @@ export const createUserReferral = async ({
...
-      data: {
-        userId: id,
-        source,
-        userReferralCodeId: referralCode?.id ?? undefined,
-      },
-    });
+    const referralCode = !!userReferralCode
+      ? await dbRead.userReferralCode.findFirst({
+          where: { code: userReferralCode },
+        })
```
Comment (manuelurenah) [inline src/server/services/user.service.ts:747]: If we are using `findFirst`, then we should add the condition to check if it's not deleted I think
```suggestion
      ? await dbRead.userReferralCode.findFirst({
          where: { code: userReferralCode, deletedAt: null },
        })
```

File: src/server/services/user.service.ts [hunk 9]
```diff
@@ -732,26 +732,44 @@ export const createUserReferral = async ({
...
+    }
+
+    if (user.referral && referralCode) {
+      // Allow to update a referral with a user-referral-code:
+      return await dbWrite.userReferral.update({
+        where: {
+          id: user.referral.id,
+        },
+        data: {
+          userReferralCodeId: referralCode?.id ?? undefined,
```
Comment (manuelurenah) [inline src/server/services/user.service.ts:761]: I believe this is not necessary anymore
```suggestion
          userReferralCodeId: referralCode.id,
```

File: src/types/next-auth.d.ts [hunk 10]
```diff
@@ -9,6 +9,7 @@ interface ExtendedUser {
   image?: string;
   email?: string;
   createdAt?: Date;
+  deletedAt?: Date;
```
Comment (manuelurenah) [inline src/types/next-auth.d.ts:12]: Might be a leftover, I don't think we should be returning the `deletedAt` field for a user session 😛 If the user is soft deleted, then there shouldn't be a session I guess?

**Pull request merged**